src/mono/mono/mini/mini-llvm.c

```diff
@@ -681,11 +685,17 @@ create_llvm_type_for_type (MonoLLVMModule *module, MonoClass *klass)
 		for (i = 0; i < size; ++i)
 			eltypes [i] = esize == 4 ? LLVMFloatType () : LLVMDoubleType ();
 	} else {
-		size = get_vtype_size (t);
-
-		eltypes = g_new (LLVMTypeRef, size);
-		for (i = 0; i < size; ++i)
-			eltypes [i] = LLVMInt8Type ();
+		MonoSizeAlign size_align = get_vtype_size_align (t);
+		eltypes = g_new (LLVMTypeRef, size_align.size);
+		size = 0;
+		int bytes = 0;
+		int chunk = size_align.align < TARGET_SIZEOF_VOID_P ? size_align.align : TARGET_SIZEOF_VOID_P;
+		for (; chunk > 0; chunk = chunk >> 1) {
+			for (; (bytes + chunk) <= size_align.size; bytes += chunk) {
+				eltypes [size] = LLVMIntType (chunk * 8);
```

Comment:
@imhameed, I'm curious, why do we use these element types that don't have anything to do with the actual elements of the original Mono type (and presumably we bitcast at the use sites?).  Why not use the field layout of the Mono type to create the LLVM struct?  (ie: for a mono class `int x; double y; represent it as `int32 x; int32 padding; double y` instead of I think `int32 el0; int32 el1; int32 el2; int32 el3`)?

Is it to limit the number of unique llvm types that we have to create?

Reply:
We were only doing it for simplicity i think.

Reply:
> @imhameed, I'm curious, why do we use these element types that don't have anything to do with the actual elements of the original Mono type (and presumably we bitcast at the use sites?). Why not use the field layout of the Mono type to create the LLVM struct? (ie: for a mono class `int x; double y;` represent it as `int32 x; int32 padding; double y` instead of I think `int32 el0; int32 el1; int32 el2; int32 el3`)?

Simplicity, mostly. The diff is small. I'd like to more accurately represent IL types in our LLVM IR in the future.

Reads and writes of individual fields are done via normal loads/stores, pointer coercion, and getelementpointer.